content/en/platform/corda/4.12/community/comm-upgrade-guide.md

```diff
@@ -0,0 +1,275 @@
+---
+aliases:
+- /head/comm-upgrade-guide.html
+- /HEAD/comm-upgrade-guide.html
+- /comm-upgrade-guide.html
+date: '2024-06-05'
+menu:
+  corda-community-4-12:
+    identifier: corda-community-4-12-comm-upgrade-guide
+    parent: corda-community-4-12-upgrading
+    weight: 15
+tags:
+- upgrade
+- guide
+title: Corda Community Edition 4.11 to 4.12 upgrade guide
+---
+
+# Corda Community Edition 4.11 to 4.12 upgrade guide
+
+This upgrade guide outlines the steps for migrating your Corda 4.11 node to version 4.12 while maintaining backwards compatibility with earlier Corda 4.x versions.
+
+{{< note >}}
+The steps from this guide only work for direct upgrades from Corda 4.11 to 4.12. If you have any nodes on versions 4.10 and below, you must upgrade them to 4.11 first. To do that, consult the relevant release upgrade documentation.
+{{< /note >}}
+
+### Background
+
+Corda 4.12 has been upgraded to use Kotlin 1.9 and Java 17. When designing Corda 4.12, we had to take into consideration the issue of having a backchain of transactions that had been verified using Kotlin 1.2 and Java 8 by earlier versions of Corda.
+
+If a Corda 4.12 node were to be part of a network of mixed Corda 4.x versions, a Corda 4.12 node would not be able to verify a contract compiled using Kotlin 1.2. Similarly, a Corda node running an earlier version of Kotlin would not be able to verify a new Corda 4.12 contract compiled using Kotlin 1.9.
+
+### The solution
+
+To address verification issues, both earlier versions of Corda 4.x and Corda 4.12 had to have a way of verifying the correct contract attachment, depending on the Corda version. This is where the external verifier, a component of Corda 4.12, comes in.
+
+The external verifier is a process started by Corda 4.12, running Kotlin 1.2 (with Java 17). This enables Corda 4.12 to run independently of Java 8. Whenever Corda 4.12 detects an old contract version, it externally verifies this contract within the external verifier process.
+
+When upgrading to Corda 4.12, the old contract CorDapp JAR is preserved and relocated to a new directory named `legacy-contracts`. This directory is essential as it provides legacy contracts for Corda 4.12 nodes to maintain backward compatibility, as required by the external verifier.
+
+To ensure compatibility of all existing transactions in the node’s backchain with the external verifier, and consequently with Corda 4.12, a new utility tool has been introduced: the Transaction Validator Utility. This tool runs the external verifier process on the existing backchain as a sanity check.
```

Comment:
The TVU is not available for Corda Community. For a community upgrade, users/operators need to figure out how to validate their existing transactions for themselves before upgrading.

Reply:
Removed.